### PR TITLE
SMS: Add Garland-Heckbert simplification

### LINKED ISSUE #3835: Add new simplification policy
We are missing a policy that simplify as much as possible according to a global error.

Possible references:
- [Surface Simplification Using Quadric Error Metrics](https://www.ri.cmu.edu/pub_files/pub2/garland_michael_1997_1/garland_michael_1997_1.pdf)
- Hugues Hoppe (all his code is available under github is we want to give it a try)
- ...?

### PR DESCRIPTION
**Pull request by MaelRL**
## Summary of Changes

This PR adds a new duo of cost and placement oracles to the Surface Mesh Simplification package, based upon the paper of Garland && Heckbert and their concept of quadric errors. Initial work by @baskinburak, during the 2019 Season of the Google Summer of Code (https://cgal.geometryfactory.com/CGAL/Members/wiki/GSoC2019/Simplification).

In addition, this PR:
- Completely cleans the package (no more Camel case, incomprehensible variable names, weird files, useless `_impl.h`, subdirectory /Details/, wonky indentation, needlessly complicated assertion system, etc.).
- Upgrade to C++11 / BGL functions.
- Properly (un)document everything.
- Fixes the concept and models: `EdgeProfile` is not a real concept, it is hard coded in the class `Edge_collapse`.
- Fixes the absence of a traits class, and thus the fact you couldn't actually use a custom VPM.

No breaking changes other than fixing the documentation to reflect the code.

Not done: fixing LT_core to properly use Eigen::Matrix and geom traits.

TODO:
- [x] Clean branch
- [x] Check doc
- [x] fix "It returns an optional value, which can be absent, in which case the remaining vertex is kept in its place."
- [x] Small feature
- [x] Check that the Eigen requirement is clear
- [x] read full diff
- [x] Benchmarks / Vtune
- [ ] news entry (?)

## Release Management

* Affected package(s): `Surface_mesh_simplification`
* Issue(s) solved (if any): Fixes #3835
* Feature/Small Feature (if any): [GH_Simplification](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/GH_Simplification) -- **pre-approval date: 2019/10/21**

### CODE REVIEW

**Comment by MaelRL** [comment]
9b30a55 and 64ba5fa change the documentation to clarify that a null placement mean no collapse should be done.

In theory, a null placement meant you keep the position of the target. But it was broken if you did not use it with a cost that is not immediately null when the placement is null (e.g. edge length cost). That was "fixed" in PR https://github.com/CGAL/cgal/pull/4160 but that patch did not follow the documentation (it should have set the placement to the target of the edge instead).

Additionally, the policy `Bounded_normal_change_placement` also goes against documentation as it aims to prevent a collapse by returning null, which, if the code was proper, would not have prevented any collapse. It should have been a wrapper around Cost policies (more details in issue https://github.com/CGAL/cgal/issues/3829).

So, there are now quite a few things going against the documentation and it would be painful to revert them. On the other hand, what was advertised by the documentation never worked and thus was never used, and is of limited utility.

Thus, I'm aligning the documentation with the behavior of the code with these two commits.

**Comment by MaelRL** [comment]
I left some new files with license text as a test. Travis should be hopefully be unhappy.

File: Surface_mesh_simplification/doc/Surface_mesh_simplification/PackageDescription.txt [hunk 1]
```diff
@@ -3,12 +3,14 @@
 /// \ingroup PkgSurfaceMeshSimplificationRef
 /*!
 \addtogroup PkgSurfaceMeshSimplificationRef
-\todo check generated documentation
 \cgalPkgDescriptionBegin{Triangulated Surface Mesh Simplification,PkgSurfaceMeshSimplification}
 \cgalPkgPicture{SMS-detail.png}
 \cgalPkgSummaryBegin
-\cgalPkgAuthor{Fernando Cacciola}
-\cgalPkgDesc{This package provides an algorithm to simplify a triangulated surface mesh  by edge collapsing.  It is an implementation of the Turk/Lindstrom <I>memoryless</I>  surface mesh simplification algorithm.}
+\cgalPkgAuthor{Fernando Cacciola, Mael Rouxel-Labbé, and Baskın Şenbaşlar}
```
Comment (sloriot) [inline Surface_mesh_simplification/doc/Surface_mesh_simplification/PackageDescription.txt]: Ş is generating a doc warning:
`WARNING: a non ascii character has been found in author string for bibtex (probably a non-handled accentuated letter).Check the new package added and update the function protect_accentuated_letters in Documentation/scripts/generate_how_to_cite.py`

You need to update this python script to handle the lower and upper case latter.

**Comment by maxGimeno** [comment]
warnings : https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-11.shtml#Surface_mesh_simplification_Examples

**Comment by MaelRL** [comment]
Segfault to investigate there: https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.1-Ic-16/Surface_mesh_simplification_Examples/TestReport_lrineau_ArchLinux-clang-Release.gz

File: Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/internal/GarlandHeckbert_core.h [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+                           CGAL::square(plane.c()));
+      const FT den = FT(1) / norm;
+
+      plane_mtr << den * plane.a(),
+                   den * plane.b(),
+                   den * plane.c(),
+                   den * plane.d();
+      quadric += plane_mtr.transpose() * plane_mtr;
+
+      if(is_discontinuity_edge(hd, tmesh))
```
Comment (sloriot) [inline Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/internal/GarlandHeckbert_core.h]: When I have twice the same code for two opposite halfedges, I usually do a for-loop(i=0;i<2) and update hd to its opposite

File: Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/internal/GarlandHeckbert_core.h [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+      {
+        const vertex_descriptor source_vd = source(hd, tmesh);
+
+        const Vector_3 p1p2 = gt.construct_vector_3_object()(get(vpm, source_vd), get(vpm, target_vd));
+        const Vector_3 normal = gt.construct_cross_product_vector_3_object()(
+                                  p1p2, gt.construct_orthogonal_vector_3_object()(plane));
+
+        const FT d = - normal * target_vertex_vector;
+        const FT norm = sqrt(CGAL::square(normal.x()) +
```
Comment (sloriot) [inline Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/internal/GarlandHeckbert_core.h]: you could use the member function

File: Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/internal/GarlandHeckbert_core.h [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+      const face_descriptor fd = face(hd, tmesh);
+      if(fd == GraphTraits::null_face())
+        continue;
+
+      Plane_3 plane(get(vpm, source(hd, tmesh)),
+                    get(vpm, target(hd, tmesh)),
+                    get(vpm, target(next(hd, tmesh), tmesh)));
+
+      Row4 plane_mtr;
+      const FT norm = sqrt(CGAL::square(plane.a()) +
```
Comment (sloriot) [inline Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/internal/GarlandHeckbert_core.h]: use member functions?

File: Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/internal/GarlandHeckbert_core.h [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+  template <typename VCM>
+  static void fundamental_error_quadrics(VCM& vcm, // quadrics container
+                                         const Triangle_mesh& tmesh,
+                                         const Vertex_point_pmap& vpm,
+                                         const Geom_traits& gt,
+                                         const FT discontinuity_multiplier = FT(100))
+  {
+    Matrix4x4 nq = Matrix4x4::Zero();
+    for(vertex_descriptor v : vertices(tmesh))
+      put(vcm, v, nq); // @todo is that necessary ?
```
Comment (sloriot) [inline Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/internal/GarlandHeckbert_core.h]: left todo

File: Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/internal/GarlandHeckbert_core.h [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+
+    for(face_descriptor f : faces(tmesh))
+    {
+      const halfedge_descriptor h = halfedge(f, tmesh);
+
+      const Point_reference p = get(vpm, source(h, tmesh));
+      const Point_reference q = get(vpm, target(h, tmesh));
+      const Point_reference r = get(vpm, target(next(h, tmesh), tmesh));
+
+      const FT rpx = p.x() - r.x();
```
Comment (sloriot) [inline Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/internal/GarlandHeckbert_core.h]: use `CGAL::determinant`?

File: Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/internal/Common.h [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+#include <set>
+
+namespace CGAL {
+namespace Surface_mesh_simplification {
+
+template<class Handle>
+inline bool handle_assigned(Handle h) { Handle null; return h != null; }
+
+template<class Iterator, class Handle>
+bool handle_exists(Iterator begin, Iterator end, Handle h)
```
Comment (sloriot) [inline Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/internal/Common.h]: is this ever used now that we have descriptors?

File: Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/internal/Common.h [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+#include <utility>
+#include <vector>
+#include <vector>
+#include <set>
+
+namespace CGAL {
+namespace Surface_mesh_simplification {
+
+template<class Handle>
+inline bool handle_assigned(Handle h) { Handle null; return h != null; }
```
Comment (sloriot) [inline Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/internal/Common.h]: should be removed as invalid (`null_halfedge()`) should be used.

File: Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/internal/GarlandHeckbert_core.h [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+  {
+    return aFirst + aSecond;
+  }
+
+  /**
+  * Returns `true` if the target of h is a discontinuity vertex in tmesh
+  *
+  * Currently, only checks if vertex belongs to a border.
+  */
+  static bool is_discontinuity_vertex(const halfedge_descriptor h, const Triangle_mesh& tmesh)
```
[the remark this answers is not in the document]
Comment (MaelRL) [reply Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/internal/GarlandHeckbert_core.h]: There is a switch that uses this function, commented farther down. I need to test/implement it further, but maybe as a later enhancement.

File: Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/internal/Common.h (hunk 7 above)
Comment (MaelRL) [reply Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/internal/Common.h]: It is used in some assertions. I marked them as expensive and moved on. Could be removed, I don't care much for them...

File: Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/internal/Common.h (hunk 8 above)
Comment (MaelRL) [reply Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/internal/Common.h]: These are not all halfedges but also vertices, edges, etc. which is why it was kept. Could just be removed if `handle_exists` is removed.

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-35.shtml

**Comment by sloriot** [comment]
@MaelRL the TODO after merge remains.

**Comment by lrineau** [comment]
[**@sloriot**](https://github.com/sloriot) commented on [Dec 20, 2019, 9:37 AM GMT+1](https://github.com/CGAL/cgal/pull/4232#issuecomment-567841278 "2019-12-20T08:37:37Z - Replied by Github Reply Comments"):
> @MaelRL the TODO after merge remains.

What is it about? It is about the news entry? It does not seem to be added to the CGAL website.

I am wondering what we could do to better track those "TODO after merge". Should we create new issue tickets?

**Comment by MaelRL** [comment]
Yes it's the new entry. I have it on my todo list, I just didn't want to do it during the holidays as get maximal exposure :]

The flag pretty much means "news entry" at the moment. I track manually and poke people when they need to do news entries. I don't think it's used enough yet to justify a systemic approach to track them.

**Pull request merged**